### PR TITLE
New: Allow regexes in RuleTester (fixes #7837)

### LINKED ISSUE #7837: Allow regexes in RuleTester
* **ESLint Version:** 3.5.0

It would be nice for the RuleTester to allow regular expressions to match messages, rather than just strings. At Facebook, some of our lint rules have very long messages - Here's one such message from a lint rule that has several possible messages:
```js
context.report({
  node: node,
  message:
    'This is a shared module but it is referencing a ' +
    '{{dependencyProduct}} module ({{requiredModule}}). Shared ' +
    'modules should not have any dependencies on product code. ' +
    'If this module is specific to {{dependencyProduct}}, it should ' +
    'be moved to {{productPath}}. Contact danlo@ if this is a ' +
    'false positive.',
  data: {
    dependencyProduct: dependencyProduct,
    productPath: getPathForProductName(dependencyProduct),
    requiredModule: requiredModule,
  },
});
```

And an example of one of the test cases:
```js
{
  // Shared module requiring a Power Editor module
  code: 'require("AdsPEApplication")',
  filename: 'html/js/ads/utils/AdsBidUtils.js',
  errors: [
    'This is a shared module but it is referencing a Power Editor module ' +
    '(AdsPEApplication). Shared modules should not have any dependencies ' +
    'on product code. If this module is specific to Power Editor, it ' +
    'should be moved to html/js/ads/powereditor. Contact danlo@ if this ' +
    'is a false positive.',
  ],
},
```

All I really want to check is that the correct message is being returned, and the correct variables are merged into it. Right now I need to copy and paste the entire message for every test case. If the message changes at all, I need to remember to update every test case. Alternatively, I could export the message from the lint rule module and import it into the unit test, but that feels a bit messy and I'd still need to deal with the variables in the string.

It would be nicer to be able to use a regex, so I could do something like:
```
  errors: [
    /Power Editor module \(AdsPEApplication\).+should be moved to html\/js\/ads\/powereditor/,
  ],
```

Alternatively, it would also be useful to be able to provide a "message ID" to the `context.report` call:
```js
context.report({
  node: node,
  message: 'Hello world',
  message_id: 2, // Any arbitrary integer
});
```

Such that the RuleTester could match the message by ID rather than by actual message text:
```js
{
  code: '// Foo bar',
  errors: [
    {id: 2}, // "2" matches the integer used in context.report
  ],
}
```

### PR DESCRIPTION
**Pull request by Daniel15**
**What is the purpose of this pull request? (put an "X" next to item)**

[x] Add something to the core

**What changes did you make? (Give an overview)**
Updated RuleTester to allow the use of regular expressions for matching rule messages, instead of just plain strings

Fixes #7837

### CODE REVIEW

**Comment by Daniel15** [comment]
I got one test failure when running unit tests locally, but it doesn't seem related to my changes and I get it even on a clean checkout of master:
```
14589 passing (1m)
1 failing

1) bin/eslint.js reading from stdin gives a detailed error message if no config file is found:
   AssertionError: expected '\nOops! Something went wrong! :(\n\nESLint couldn\'t find the plugin "eslint-plugin-react". This can happen for a couple different reasons:\r\n\r\n1. If ESLint is installed globally, then make sure eslint-plugin-react is also installed globally. A globally-installed ESLint cannot find a locally-installed plugin.\r\n\r\n2. If ESLint is installed locally, then it\'s likely that the plugin isn\'t installed correctly. Try reinstalling by running the following:\r\n\r\n    npm i eslint-plugin-react@latest --save-dev\r\n\r\nIf you still can\'t figure out the problem, please stop by https://gitter.im/eslint/eslint to chat with the team.\r\n\n' to match /ESLint couldn't find a configuration file/
    at Function.assert.match (C:\src\eslint\node_modules\chai\lib\chai\interface\assert.js:883:32)
    at getStdout.then.stdout (C:\src\eslint\tests\bin\eslint.js:92:24)
    at process._tickCallback (internal/process/next_tick.js:103:7)
```

**Comment by platinumazure** [comment]
@Daniel15 `npm install` from latest master should fix that error for you.

File: tests/lib/testers/rule-tester.js [hunk 1]
```diff
@@ -122,6 +122,17 @@ describe("RuleTester", () => {
...
+            ruleTester.run("no-var", require("../../fixtures/testers/rule-tester/no-var"), {
+                valid: [],
+                invalid: [
+                    { code: "var foo = bar;", errors: [{ message: /Bad error message/ }] }
+                ]
+            });
+        }, /Expected 'Bad var.' to match \/Bad error message\//);
+    });
+
     it("should throw an error when the error is neither an object nor a string", () => {
```
Comment (alberto) [inline tests/lib/testers/rule-tester.js]: Can you update the description of the test accordingly?

Comment (Daniel15) [inline tests/lib/testers/rule-tester.js]: @alberto - I changed the description to "should throw an error when the error is not a supported type", let me know if that's okay.

**Pull request merged**